### ISSUE TITLE
Istio Operator does not uninstall components

### ISSUE THREAD

**Issue by georgiiturevich-uw** [istio/istio#35016]
### Bug Description

**In short**:
Istio Operator does not uninstal Istio resources when I delete an IstioOperator CR. It constantly reporting `Reconciling IstioOperator` and `Deleting IstioOperator istio-control-plane-1-11-2`.

**Setup**:
- Created cluster with `kind` utility
```
kind create cluster --name istio-test --config ./kind_istio-test-1.20.yaml
```
where `kind_istio-test-1.20.yaml` containes the following
```
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  image: kindest/node:v1.20.2@sha256:8f7ea6e7642c0da54f04a7ee10431549c0257315b3a634f6ef2fecaaedb19bab
- role: worker
  image: kindest/node:v1.20.2@sha256:8f7ea6e7642c0da54f04a7ee10431549c0257315b3a634f6ef2fecaaedb19bab
```
- Installed Istio Operator
```
./istio-1.11.2/bin/istioctl operator init --watchedNamespaces=istio-system --revision=1-11-2
```
- Installed Control Plane and Gateways IstioOperator CRs
```
kubectl apply -f - <<EOF
---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istio-control-plane-1-11-2
spec:
  revision: "1-11-2"
  profile: minimal
EOF


kubectl apply -f - <<EOF
---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: gateways
spec:
  profile: empty
  revision: "1-11-2"
  components:
    egressGateways:
      - name: istio-egressgateway
        enabled: true
    ingressGateways:
      - name: istio-ingressgateway
        enabled: true
EOF
```
- As a result I saw istiod and gateways pods in `istio-system` namespace
```
$ kubectl get pods -n istio-system
NAME                                    READY   STATUS    RESTARTS   AGE
istio-egressgateway-698f9dc949-5lxbg    1/1     Running   0          37m
istio-ingressgateway-55d9556456-xnpdh   1/1     Running   0          37m
istiod-1-11-2-64d6c964b9-27w8l          1/1     Running   0          37m
```
** IstioOperator CRs deletion **
When I try to remove one of this CRs then `kubectl` hangs
```
$ kubectl delete istiooperators.install.istio.io -n istio-system istio-control-plane-1-11-2 
istiooperator.install.istio.io "istio-control-plane-1-11-2" deleted

```
In fact the CR and all related Kubernetes resources are not deleted.
During this time I see constantly repeating `Reconciling IstioOperator` and `Deleting IstioOperator istio-control-plane-1-11-2` messages in istio-operator pod logs. For example:
```
...
2021-09-03T15:34:41.765435Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
2021-09-03T15:34:42.098423Z	info	installer	Reconciling IstioOperator
2021-09-03T15:34:42.098804Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
2021-09-03T15:34:42.753390Z	info	installer	Reconciling IstioOperator
2021-09-03T15:34:42.753862Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
2021-09-03T15:34:44.047766Z	info	installer	Reconciling IstioOperator
2021-09-03T15:34:44.048132Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
2021-09-03T15:34:46.632213Z	info	installer	Reconciling IstioOperator
2021-09-03T15:34:46.632669Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
2021-09-03T15:34:51.768007Z	info	installer	Reconciling IstioOperator
2021-09-03T15:34:51.768424Z	info	installer	Deleting IstioOperator istio-control-plane-1-11-2
...
```
With Istio 1.10.4 I also can not uninstall components. `kubectl` can successfully delete IstioOperator CRs and it does not hang, but related Kubernetes resources still exist and the following messages appear in istio-operator pod logs
```
2021-09-03T16:08:21.051523Z	info	installer	Reconciling IstioOperator
2021-09-03T16:08:21.051771Z	info	installer	Deleting IstioOperator istio-control-plane-1-10-4
2021-09-03T16:08:21.070348Z	info	installer	API server conflict during finalizer removal, retrying.
2021-09-03T16:08:21.078207Z	info	installer	Reconciling IstioOperator
```

### Version

```prose
$ istioctl version
client version: 1.11.2
control plane version: 1.11.2
data plane version: 1.11.2 (2 proxies)
```
```
$ kubectl version --short
Client Version: v1.19.7
Server Version: v1.20.2
```
The same behavior is when server version is v1.19.7

```
$ kind version
kind v0.10.0 go1.15.7 linux/amd64
```

### Additional Information

[bug-report.zip](https://github.com/istio/istio/files/7107171/bug-report.zip)

**Comment by georgiiturevich-uw** [comment]
It may be a duplicate of https://github.com/istio/istio/issues/33649

**Comment by xichengliudui** [comment]
![WechatIMG6505](https://user-images.githubusercontent.com/40875627/132275688-f0b01d3b-5164-4174-8f34-80e811c3f1a6.png)

**Comment by xichengliudui** [comment]
check whether you have pods/proxy resource permissions
![image](https://user-images.githubusercontent.com/40875627/132275892-931ec1b5-5f57-4da4-aed6-fd974d4df77d.png)

**Comment by georgiiturevich-uw** [comment]
@xichengliudui Hello Liu, yes, I have this permission.

**Comment by georgiiturevich-uw** [comment]
If I do not use `revision` during Istio Operator installation and IstioOperator CRs do not have 'revision' attribute.
```
./istio-1.11.2/bin/istioctl operator init --watchedNamespaces=istio-system

kubectl apply -f - <<EOF
---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istio-control-plane-1-11-2
spec:
  profile: minimal
EOF


kubectl apply -f - <<EOF
---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: gateways
spec:
  profile: empty
  components:
    egressGateways:
      - name: istio-egressgateway
        enabled: true
    ingressGateways:
      - name: istio-ingressgateway
        enabled: true
EOF
```

then everything works as expected. So probably the issue with 'revision' based installation.

**Comment by litong01** [comment]
@georgiiturevich-uw The problem is that the serviceaccount for the operator lacks the capability of pods/portforward. When trying to deal the deletion with revision, operator calls GetIDsFromProxyInfo which calls GetProxyInfo, GetProxyInfo then calls kubeClient.AllDiscoveryDo which apparently needs pods/portforward, otherwise, that method throws exception when revision is on, the reconciler sadly tries again and again, and kubectl never returns. That is the reason why it hangs when delete the resource. The easiest way to fix this issue for your current cluster is to add pods/portforward to your istio-operator clusterrole, in your case, the role name should be istio-operator-1-11-2. A new PR will be coming up.

**Comment by litong01** [comment]
@georgiiturevich-uw please test with the fix and close this issue. Thanks.

**Comment by georgiiturevich-uw** [comment]
Hello @litong01 , sorry for the delay.

Thank you for trying fixing this!

I tried your fix.

The behavior changed a bit. `kubectl` did not hang when I tried to delete on of the IstioOperator CRs `kubectl delete io -n istio-system gateways`, and this CR was deleted, so I do not see it in the list of istiooperators. But gateways pods are still here and I see the following in istio-operator-*** pod logs
```
2021-09-29T12:35:18.589273Z	info	installer	Reconciling IstioOperator
2021-09-29T12:35:18.589880Z	info	installer	Deleting IstioOperator gateways
2021-09-29T12:35:18.675358Z	info	installer	API server conflict during finalizer removal, retrying.
2021-09-29T12:35:18.682498Z	info	installer	Reconciling IstioOperator
2021-09-29T12:35:18.682699Z	info	installer	Reconciling IstioOperator
```
It is similar to 1.10.4 behavior which I also described in the issue.

**Comment by strantalis** [comment]
I am experiencing the same issue and doesn't seem like there is much more to go off of even when debug logging is enabled.

**Comment by strantalis** [comment]
I could be wrong but I think I found why its not cleaning up the resources at least in my case. 

This is my `IstioOperator` definition
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: ingressgateway-1-10-3
spec:
  profile: empty
  revision: 1-10-3
  components:
    ingressGateways:
      - enabled: true
        name: ingressgateway-1-10-3
        k8s:
          serviceAnnotations:
            # service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
            service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: "instance"
            service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
            service.beta.kubernetes.io/aws-load-balancer-type: external
```

After debugging the operator locally it seems like the operator sets a ACTION_REQUIRED status but goes onto remove the finalizer without really checking the status [here](https://github.com/istio/istio/blob/master/operator/pkg/controller/istiocontrolplane/istiocontrolplane_controller.go#L255).

Not sure if `Delete()` should return an error or something else but because the `IstioOperator` resource gets cleaned up there is no way for us to see that status.

https://github.com/istio/istio/blob/master/operator/pkg/controller/istiocontrolplane/istiocontrolplane_controller.go#L255 

```
status:
  message: 'there are proxies still pointing to the pruned control plane: reviews-v1-545db77b95-9bdz4.default
    ingressgateway-1-10-3-749fd9b8c7-sz6c4.istio-system productpage-v1-6b746f74dc-rd7fp.default
    details-v1-79f774bdb9-zxx8j.default reviews-v2-7bf8c9648f-pfdmv.default ratings-v1-b6994bb9-q6gsf.default
    reviews-v3-84779c7bbc-p5wbs.default.'
  status: ACTION_REQUIRED
```

**Comment by morvencao** [comment]
fixed in https://github.com/istio/istio/pull/35208 and https://github.com/istio/istio/pull/35651

**Comment by jrsdav** [comment]
Just to note, the change in #35208 is not reflected in operator helm chart for the 1.11.4 release. 

I'm not sure if this is expected by the maintainers, but for those using the [istio-operator Helm chart](https://github.com/istio/istio/tree/9f6f03276054bb62a1b745630322314ef14969e8/manifests/charts/istio-operator)  you'll need to add `pods/portforward` permissions to the [ClusterRole](https://github.com/istio/istio/blob/9f6f03276054bb62a1b745630322314ef14969e8/manifests/charts/istio-operator/templates/clusterrole.yaml#L102) in order to fix this issue.

**Comment by morvencao** [comment]
@jrsdav The [fix](https://github.com/istio/istio/pull/35208) was cherrypicked in to release-1.11 in https://github.com/istio/istio/pull/35651, it will be available in 1.11.5.

**Comment by gvinayakk** [comment]
One another issue that I could see it dint clean up the validatorwebhookconfiguration.

#####
apiVersion: admissionregistration.k8s.io/v1
kind: ValidatingWebhookConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"admissionregistration.k8s.io/v1","kind":"ValidatingWebhookConfiguration","metadata":{"annotations":{},"labels":{"app":"istiod","install.operator.istio.io/owning-resource":"control-plane-1-11-4","install.operator.istio.io/owning-resource-namespace":"istio-system","istio":"istiod","istio.io/rev":"1-11-4","operator.istio.io/component":"Pilot","operator.istio.io/managed":"Reconcile","operator.istio.io/version":"1.11.4","release":"istio"},"name":"istio-validator-1-11-4-istio-system"},"webhooks":[{"admissionReviewVersions":["v1beta1","v1"],"clientConfig":{"caBundle":"","service":{"name":"istiod-1-11-4","namespace":"istio-system","path":"/validate"}},"failurePolicy":"Ignore","name":"rev.validation.istio.io","objectSelector":{"matchExpressions":[{"key":"istio.io/rev","operator":"In","values":["1-11-4"]}]},"rules":[{"apiGroups":["security.istio.io","networking.istio.io"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["*"]}],"sideEffects":"None"},{"admissionReviewVersions":["v1beta1","v1"],"clientConfig":{"caBundle":"","service":{"name":"istiod-1-11-4","namespace":"istio-system","path":"/validate"}},"failurePolicy":"Ignore","name":"validation.istio.io","objectSelector":{"matchExpressions":[{"key":"istio.io/rev","operator":"DoesNotExist"}]},"rules":[{"apiGroups":["security.istio.io","networking.istio.io","telemetry.istio.io"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["*"]}],"sideEffects":"None"}]}
  creationTimestamp: "2021-11-23T15:48:34Z"
  generation: 20
  labels:
    app: istiod
    install.operator.istio.io/owning-resource: control-plane-1-11-4
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio: istiod
    istio.io/rev: 1-11-4
    operator.istio.io/component: Pilot
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.11.4
    release: istio
  name: istio-validator-1-11-4-istio-system
  resourceVersion: "328373590"
  selfLink: /apis/admissionregistration.k8s.io/v1/validatingwebhookconfigurations/istio-validator-1-11-4-istio-system
####
